Quotient/e2ee/cryptoutils.cpp

```diff
@@ -0,0 +1,256 @@
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+    return output;
+}
+
+Expected<QByteArray, uint64_t> Quotient::aesCtr256Decrypt(const QByteArray& ciphertext, const QByteArray& aes256Key, const QByteArray& iv)
+{
+    auto context = EVP_CIPHER_CTX_new();
+    Q_ASSERT(context);
+
+    int length = 0;
+    int plaintextLength = 0;
+    QByteArray decrypted(ciphertext.size(), u'\0');
+
+    int status = EVP_DecryptInit_ex(context, EVP_aes_256_ctr(), nullptr, reinterpret_cast<const unsigned char *>(aes256Key.data()), reinterpret_cast<const unsigned char *>(iv.data()));
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+
+    status = EVP_DecryptUpdate(context, reinterpret_cast<unsigned char*>(decrypted.data()), &length, reinterpret_cast<const unsigned char*>(&ciphertext.data()[0]), (int) ciphertext.size());
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+
+    plaintextLength = length;
+    status = EVP_DecryptFinal_ex(context, reinterpret_cast<unsigned char*>(decrypted.data()) + length, &length);
+    if (status != 1) {
+        qWarning() << ERR_error_string(ERR_get_error(), nullptr);
+        return ERR_get_error();
+    }
+
+    plaintextLength += length;
+    decrypted.resize(plaintextLength);
+    EVP_CIPHER_CTX_free(context);
+    return decrypted;
+}
+
+Expected<QByteArray, uint64_t> Quotient::curve25519AesSha2Decrypt(QByteArray ciphertext, const QByteArray &privateKey, const QByteArray &ephemeral, const QByteArray &mac)
+{
+    auto context = makeCStruct(olm_pk_decryption, olm_pk_decryption_size, olm_clear_pk_decryption);
+    Q_ASSERT(context);
+
+    QByteArray publicKey(olm_pk_key_length(), u'\0');
+    auto status = olm_pk_key_from_private(context.get(), publicKey.data(), publicKey.size(), privateKey.data(), privateKey.size());
+    if (status == olm_error()) {
+        return olm_pk_decryption_last_error_code(context.get());
+    }
+
+    QByteArray plaintext(olm_pk_max_plaintext_length(context.get(), ciphertext.size()), u'\0');
+    auto result = olm_pk_decrypt(context.get(), ephemeral.data(), ephemeral.size(), mac.data(), mac.size(), ciphertext.data(), ciphertext.size(), plaintext.data(), plaintext.size());
+
+    if (result == olm_error()) {
+        return olm_pk_decryption_last_error_code(context.get());
+    }
+    plaintext.resize(result);
+    return plaintext;
+}
+
+Expected<Curve25519Encrypted, uint64_t> Quotient::curve25519AesSha2Encrypt(const QByteArray& plaintext, const QByteArray& publicKey)
+{
+    auto context = makeCStruct(olm_pk_encryption, olm_pk_encryption_size, olm_clear_pk_encryption);
+
+    auto status = olm_pk_encryption_set_recipient_key(context.get(), publicKey.data(), publicKey.size());
+    if (status == olm_error()) {
+        return olm_pk_encryption_last_error_code(context.get());
```

Comment:
Took me sometime to ~~figure out~~ remember why MSVC fails to compile this.

Welcome to the small lovely world of implementation-defined underlying types. In a nutshell, because `enum OlmErrorCode` doesn't specify its underlying type GCC and Clang assume it to be unsigned while MSVC thinks it should be signed. None of the opinions is wrong, the standard doesn't mandate either. So, when you compile this code with MSVC the compiler doesn't want to convert `OlmErrorCode` to an unsigned type without an explicit cast.

Given that the whole `Expected<>` construct is repeated across `cryptoutils.*`,  I suggest to add a `using` declaration for it, in the following form:
```cplusplus
template <typename T>
using OlmExpected = Expected<T, std::underlying_type_t<OlmErrorCode>>;
```
and use it throghout the codebase. HTH.

Reply:
...and I just found that we already have `QOlmExpected`... so the name for the one in this PR should probably be something less confusing.

Reply:
...and looking further into the code, I get an uneasy feeling that nothing guarantees error values from Olm and OpenSSL from clashing - so I'm afraid we just cannot pack things like that. Also, as discussed earlier here, probably we have to go through this code and see which errors are even worth returning - I suppose most of OpenSSL errors are not really recoverable and/or are effectively the same from the caller's perspective.

Reply:
I made a shot at fixing this along the lines mentioned in the chat. Didn't really like the fact that my changes didn't break anything outside of `cryptoutilus.*` - might mean having to look through all call sites and see how errors are dealt with _there_. But maybe it's fine...